e2etest/l2tests/interface_selector.go

```diff
@@ -0,0 +1,148 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2eservice "k8s.io/kubernetes/test/e2e/framework/service"
+	admissionapi "k8s.io/pod-security-admission/api"
+
+	"go.universe.tf/metallb/e2etest/pkg/executor"
+	"go.universe.tf/metallb/e2etest/pkg/k8s"
+	"go.universe.tf/metallb/e2etest/pkg/mac"
+	"go.universe.tf/metallb/e2etest/pkg/service"
+	internalconfig "go.universe.tf/metallb/internal/config"
+)
+
+var _ = ginkgo.Describe("L2-interface selector", func() {
+	var cs clientset.Interface
+
+	var f *framework.Framework
+	ginkgo.AfterEach(func() {
+		if ginkgo.CurrentGinkgoTestDescription().Failed {
+			k8s.DumpInfo(Reporter, ginkgo.CurrentGinkgoTestDescription().TestText)
+		}
+
+		// Clean previous configuration.
+		err := ConfigUpdater.Clean()
+		framework.ExpectNoError(err)
+	})
+
+	f = framework.NewDefaultFramework("l2")
+	f.NamespacePodSecurityEnforceLevel = admissionapi.LevelPrivileged
+
+	ginkgo.BeforeEach(func() {
+		if DefaultInterface == "" || ExtraInterface == "" {
+			framework.Fail("Cluster nodes don't have multi-interfaces to test L2-interface selector")
+		}
+
+		cs = f.ClientSet
+
+		ginkgo.By("Clearing any previous configuration")
+
+		err := ConfigUpdater.Clean()
+		framework.ExpectNoError(err)
+	})
+
+	ginkgo.Context("Specify interfaces connected client", func() {
+		ginkgo.It("Should work", func() {
+			resources := internalconfig.ClusterResources{
+				Pools: []metallbv1beta1.IPAddressPool{
+					{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: "l2-test",
+						},
+						Spec: metallbv1beta1.IPAddressPoolSpec{
+							Addresses: []string{
+								IPV4ServiceRange,
+								IPV6ServiceRange},
+						},
+					},
+				},
+				L2Advs: []metallbv1beta1.L2Advertisement{
+					{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: "with-interfaces",
+						},
+						Spec: metallbv1beta1.L2AdvertisementSpec{
+							Interfaces: []string{DefaultInterface},
+						},
+					},
+				},
+			}
+			err := ConfigUpdater.Update(resources)
+			framework.ExpectNoError(err)
+
+			svc, _ := service.CreateWithBackend(cs, f.Namespace.Name, "lb-service")
+			defer func() {
+				err := cs.CoreV1().Services(svc.Namespace).Delete(context.TODO(), svc.Name, metav1.DeleteOptions{})
+				framework.ExpectNoError(err)
+			}()
+			ingressIP := e2eservice.GetIngressPoint(&svc.Status.LoadBalancer.Ingress[0])
+			out, err := executor.Host.Exec("ip", "neigh", "flush", "to", ingressIP)
+			framework.ExpectNoError(err, out)
+
+			ginkgo.By("checking connectivity to its external VIP")
+			gomega.Eventually(func() error {
+				err := mac.RequestAddressResolution(ingressIP, executor.Host)
+				if err != nil {
+					return err
+				}
+				return service.ValidateL2(svc)
+			}, 1*time.Minute, 1*time.Second).Should(gomega.Not(gomega.HaveOccurred()))
+		})
+	})
+
+	ginkgo.Context("Specify interfaces which disconnect client", func() {
+		ginkgo.It("Shouldn't work", func() {
```

Comment:
Something I didn't notice before. Why using the second interface shouldn't work?
Both interfaces should be accessible from the host via their respective bridges I think.

I think what is making it not working is the fact that the ipv4service range is related to the default bridge range. I did a quick test hardcoding the ip to the one of network one and it passed, so arping looks for the subnet in order to use the interface to request the arp from.

In short, we should validate when requesting arping via both bridges (using the -I option) work only from the selected ones.
IF we want to validate wget via the secondary network, we need to know its subnet and set the iprange of the addresspool accordingly.

Does it make sense? Also cc @oribon

Reply:
I has update the e2etest. One is about mac check, another is about arp check.